src/client_side.cc

```diff
@@ -2095,7 +1912,7 @@ ConnStateData::clientParseRequests()
 
         // try to parse the PROXY protocol header magic bytes
         if (needProxyProtocolHeader_) {
-            if (!parseProxyProtocolHeader())
+            if (!parseProxyProtocolMessage())
```

Comment:
> I know a lot of Squid code uses the term "header" inconsistently and for the wrong things, but no need to follow that bad practice in the new code.

I do not think this PR uses the term "header" inconsistently and for the wrong things. You may argue otherwise, but the remainder of your change request appears to be unrelated to the above statement -- it talks about replacing "Message" with "Header" in new code, not fixing inconsistent/wrong "header" usage.


> The PROXY protocol defines a "connection header". 

The PROXY protocol specs certainly talk about a "header" at the beginning of a connection, but given the overall low quality of the PROXY protocol specs _and_ their half-baked transition from v1 (where the prefix was a lot more like a small header) to v2 (where the prefix has all the features of a full-blown message, from a "start line" to a variable-length multipart "body" with varying content types), that talk is not sufficient reason to propagate that terminology to Squid.

We decided to call "PROXY protocol header" a message because that was a better choice than the alternatives, even though the PROXY protocol specs use the term "header". It is much easier to realize (and get used to the fact) that Squid PROXY "message" is what the PROXY specs call "header", than to navigate the mess of mapping "PROXY header" to Squid APIs that use "header" for different purposes. This becomes especially clear when you think of fixed/always-there PROXY blocks such as `:command` -- they form the request line that belongs to Http::Message equivalent, not HttpHeader equivalent.

At the beginning, we were not sure whether to use "header" or "message", but I believe that subsequent development (the final result of which you see in this PR) has proven that "message" was the right choice. Its mismatch with the PROXY protocol terminology is a minor problem that is trivial to overcome. The mismatch between ProxyProtocol::Header and Http::Header (and related existing APIs) would be much uglier.

> This is very distinctly and explicitly a prefix header for the entire connection, it cannot be used like a message.

Why not?! Just because protocol X bytes are followed by protocol Y bytes, does not mean X bytes cannot be a message. For example, HTTP/1 CONNECT or Upgrade:websockets request is still a message.

> The HTTP 'header' objects which you are using internally to represent the pieces of PROXY data are formally defined by HTTP as "fields" with a field name and field value. So the types and functions for manipulating them should refer to Field, FieldName, FieldValue etc.

Where HTTP APIs are used for PROXY messages, those APIs should continue to be used with all their (poor) terminology choices. Given the volume and age of those APIs, it is much better to stay in sync with them than to fight them. I agree that HTTP APIs can be significantly improved by using field-based terminology, but if that improvement is done (and that is a big `if`!), it should be done properly across the entire code base, in a dedicated PR(s).

@yadij, I hope you do not insist on renaming ProxyProtocol::Message into ProxyProtocol::Header. Do you?

Reply:
Amos:
>  I know a lot of Squid code uses the term "header" inconsistently and for the wrong things, but no need to follow that bad practice in the new code.

Alex:
> I do not think this PR uses the term "header" inconsistently and for the wrong things. You may argue otherwise, but the remainder of your change request appears to be unrelated to the above statement -- it talks about replacing "Message" with "Header" in new code, not fixing inconsistent/wrong "header" usage.

Other code uses "header" to mean both the mime area of messages and an individual field in that. This PR currently re-names what the protocol specification defines as a "header" to be "message" and field-names to "headers". Even HTTP does not use that combination of definitions - a "message" there is a request-line or status-line and 0-N segments/frames of other stuff.
How can you say that this PR is not perpetuating the bad / vague naming practice?


Amos:
> This is very distinctly and explicitly a prefix header for the entire connection, it cannot be used like a message.

Alex:
> Why not?! Just because protocol X bytes are followed by protocol Y bytes, does not mean X bytes cannot be a message. For example, HTTP/1 CONNECT or Upgrade:websockets request is still a message.

This is exactly the type of confused conclusion that results from thinking of the PROXY header as a message. The bytes forming a PROXY protocol header cannot and must not be delivered after any previous bytes in that same connection. It is defining the properties of the native protocol(s) within which it is wrapped.

Eduard:
> a developer, unfamiliar with PROXY protocol, could mistakenly expect that sometimes there may be PROXY bytes beyond this 'header'. Naming all PROXY-protocol-related bytes as 'message' resolves such confusion.

Firstly this object is a MiME representation of the PROXY protocol. Mime headers do not have anything "following". They are by definition a list of things: the end of the list is the end of the list, with no such concept as "the list entry *after* the end of the list".

Secondly, a sane developer would look for the specifications of these bjects. Especially since we should be referencing the specification the documents in relevant object documentation. If one searches for the terms used in this PR in either the PROXY or HTTP specifications one finds text describing "message" as a thing - that text clearly indicates messages as things with multiple and repeated uses - with conflicts in a critical way with the actual PROXY definition for use of the "header" bytes here being called a "Message".

This thing which is only defined here and being called "Message" is representing the PROXY octets in the form of a "MiME header". Which is formally defined in the MiME specification (re-used by HTTP, etc) as a "set of header fields" (note that header here is part of a compound noun, much like "transparent" the second word actually has more specific meaning for concept clarity).

In short - I do not see any confusion if we refer to the field-name:field-value pairs as fields, field-name, and field-value. Those terms are consistent with the HTTP, MiME and PROXY specifications terminology for and thus should be relatively familiar to anyone. Likewise "Header" for a set of such fields is also consistent for all those specifications (if a bit loose).

Reply:
I do. On grounds that the PROXY spec document uses the term "header" as its definition for the protocol object being represented. If you want to discuss altering that naming with Willy T. go ahead, the document is technically still in draft and he is very open to improvements if we have issues.

I just want the terminology we use in the code symbol names to align with the specification texts that code derives from so we don't have to re-educate people on why things are "misnamed" or impose a particular view of the world as necessary to even read the code. We have enough problems in that regards with the English language dialects barrier between ourselves.

Reply:
> I do. On grounds that the PROXY spec document uses the term "header"

Done at b966eb and a25f68.

> Do you want s/HeaderType/FieldType/i in ProxyProtocol code? I am OK with that change.

Done at 35898f.